crates/kornia-apriltag/src/threshold.rs

```diff
@@ -0,0 +1,330 @@
+///    - If the difference between `local_max` and `local_min` is less than `min_white_black_diff`,
+///      the tile is skipped, and its pixels are marked as [PixelTrait::SKIP_PROCESSING].
+///    - Otherwise, the threshold for the tile is computed as:
+///      `threshold = local_min + (local_max - local_min) / 2`.
+///    - Pixels in the tile are binarized based on whether they are above or below the threshold.
+/// 3. Neighboring tiles are considered to refine the threshold for each tile, ensuring smooth transitions.
+///
+/// # Recommended Values for `min_white_black_diff`
+///
+/// | Image Type | `min_white_black_diff` |
+/// |------------|------------------------|
+/// | 8-bit      | 10-20                  |
+/// | 16-bit     | 500-1000               |
+/// | 32-bit     | 100,000-1,000,000      |
+/// | float      | 0.05-0.1               |
+///
+/// # Examples
+///
+/// ```
+/// use kornia_image::{allocator::CpuAllocator, Image, ImageSize};
+/// use kornia_apriltag::threshold::{adaptive_threshold, TileBuffers};
+///
+/// let src = Image::new(
+///     ImageSize {
+///         width: 2,
+///         height: 3,
+///     },
+///     vec![0, 50, 100, 150, 200, 250],
+///     CpuAllocator,
+/// )
+/// .unwrap();
+/// let mut dst = Image::from_size_val(src.size(), 0u8, CpuAllocator).unwrap();
+///
+/// let mut tile_buffers = TileBuffers::new(src.size(), 2);
+/// adaptive_threshold(&src, &mut dst, &mut tile_buffers, 2, 20).unwrap();
+/// assert_eq!(dst.as_slice(), &[0, 0, 0, 255, 255, 255]);
+/// ```
+// TODO: Add support for parallelism
+pub fn adaptive_threshold<
+    T: PixelTrait
+        + PartialOrd
+        + Copy
+        + Add<T, Output = T>
+        + Sub<T, Output = T>
+        + Div<T, Output = T>
+        + From<u8>,
+    A1: ImageAllocator,
+    A2: ImageAllocator,
+>(
+    src: &Image<T, 1, A1>,
+    dst: &mut Image<T, 1, A2>,
+    tile_buffers: &mut TileBuffers<T>,
+    tile_size: usize,
+    min_white_black_diff: T,
+) -> Result<(), AprilTagError> {
+    if src.size() != dst.size() {
+        return Err(
+            ImageError::InvalidImageSize(src.cols(), src.rows(), dst.cols(), dst.rows()).into(),
+        );
+    }
+
+    let tile_iterator = TileIterator::from_image(src, tile_size);
+    let tiles_x_len = tile_iterator.tiles_x_len();
+    let tiles_y_len = tile_iterator.tiles_y_len();
+
+    let expected_tile_count = tiles_x_len * tiles_y_len;
+    if tile_buffers.tile_min.len() != expected_tile_count {
+        // It is guaranteed for tile_min and tile_max to have same length by design
+        // so, avoiding additional check for tile_max
+        return Err(AprilTagError::InvalidTileBufferSize(
+            tile_buffers.tile_min.len(),
+            expected_tile_count,
+        ));
+    }
+
+    let dst_data = dst.as_slice_mut();
+
+    // Calculate extrema (i.e. min & max grayscale value) of each tile
+    for (i, tile) in tile_iterator.enumerate() {
```

Comment:
can we simplify first with plain iterators ? so that it's easy to expand lwater to par_iter

Reply:
i think that you can just iterate per pixel and have pre-computed offsets (as you know the image cols) to get the max/min per tile in a 3x3 neighbourhood